### ISSUE TITLE
[Flaky Test] k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler.reconciler Test_UncertainDeviceGlobalMounts

### ISSUE THREAD

**Issue by drewhagen** [kubernetes/kubernetes#127520]
### Which jobs are flaking?
* master-blocking
  * ci-kubernetes-unit

### Which tests are flaking?
`k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler.reconciler`

### Since when has it been flaking?
Flaked on September 20, 2024  
[Link to failure](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit/1837283673959829504)

### Testgrid link
[TestGrid: ci-kubernetes-unit](https://testgrid.k8s.io/sig-release-master-blocking#ci-kubernetes-unit)

### Reason for failure (if possible)
```
{Failed;Failed; === RUN Test_UncertainDeviceGlobalMounts/timed_out_operations_should_result_in_device_marked_as_uncertain_[Block][ I0921 00:32:12.806506 56545 operation_generator.go:1014] "MapVolume.WaitForAttach entering for volume "success-and-timeout-device-name"" pod="pod1" E0921 00:32:12.807049 56545 reconstruct.go:54] "Cannot get volumes from disk, skip sync states for volume reconstruction" err="open fake-dir-6ec1e204152d0da11b247a1c9de569d6: no such file or directory" --- FAIL: Test_UncertainDeviceGlobalMounts/timed_out_operations_should_result_in_device_marked_as_uncertain_[Block][ (0.60s) }
```

### Anything else we need to know?
Failed on commit 52095a8b but also passed twice - thus flaky.

#### 1.32.0-alpha.1 is due to be cut on October 1st.
Is this an issue that could block this next release cut? Please let me know. Thank you!

### Relevant SIG(s)
/sig storage  
/cc @kubernetes/release-team-release-signal
/test flake

**Comment by carlory** [comment]
I will look into it once it occurs on the `ci-kubernetes-unit` job.

![image](https://github.com/user-attachments/assets/2dfeabe5-9ba7-4780-b8cb-c6d684f26735)

**Comment by drewhagen** [comment]
@carlory @pacoxu Thanks for investigating!

The first release cut (1.32.0-alpha.1) is due in less than a week from today on Oct 1st 2024. Given that this flake is on master blocking grid but this particular issue has been green lately, can we consider this a Non-Blocker for this next release cut? Please advise - thank you!

**Comment by pacoxu** [comment]
not a blocker to me 😄

**Comment by tico88612** [comment]
Hi @pacoxu @carlory 

It's the same error, but for a different reason: `Test_Run_Positive_VolumeMountControllerAttachEnabledRace` has happened three times. 

Triage:
- https://storage.googleapis.com/k8s-triage/index.html?test=k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler.reconciler#b5d35db100f5d9d14818
- https://storage.googleapis.com/k8s-triage/index.html?test=k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler.reconciler#544b6ab1f92846135f85

Prow:
- https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit/1839779379312332800
- https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit/1836752669566308352
- https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit/1836934117132865536

```
{Failed;  === RUN   Test_Run_Positive_VolumeMountControllerAttachEnabledRace
E0927 21:49:38.148897   43402 reconstruct.go:54] "Cannot get volumes from disk, skip sync states for volume reconstruction" err="open fake-dir: no such file or directory"
...Truncate...
E0927 21:49:38.884099   43402 reconstruct.go:54] "Cannot get volumes from disk, skip sync states for volume reconstruction" err="open fake-dir: no such file or directory"
...Truncate...
E0927 21:49:48.270562   43402 nestedpendingoperations.go:348] Operation for "{volumeName:fake-plugin/fake-device1 podName: nodeName:}" failed. No retries permitted until 2024-09-27 21:49:56.270502424 +0000 UTC m=+79.723678511 (durationBeforeRetry 8s). Error: recovered from panic "close of closed channel". (err=<nil>) Call stack:
goroutine 407 [running]:
...Truncate...
    reconciler_test.go:1986: Timed out waiting for volume "fake-plugin/fake-device1" to be attached.
--- FAIL: Test_Run_Positive_VolumeMountControllerAttachEnabledRace (13.30s)
```

Would the new fail be blocker for alpha 1 cut (Oct 1st 2024) or not? Thank you!

**Comment by pacoxu** [comment]
/cc @xing-yang @jsafrane 
CC sig-storage leads

**Comment by carlory** [comment]
/unassign

> It's the same error, but for a different reason: Test_Run_Positive_VolumeMountControllerAttachEnabledRace has happened three times.

It is different from this issue, so I created a new issue for it.

**Comment by pacoxu** [comment]
> Test_UncertainDeviceGlobalMounts

https://storage.googleapis.com/k8s-triage/index.html?text=Test_UncertainDeviceGlobalMounts 
Flake only once in last 2 weeks.

**Comment by pacoxu** [comment]
/retitle [Flaky Test] k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler.reconciler Test_UncertainDeviceGlobalMounts
So we can know which dedicated flake it is.

**Comment by pacoxu** [comment]
/sig node storage 
/cc @jsafrane 
https://storage.googleapis.com/k8s-triage/index.html?text=Test_UncertainDeviceGlobalMounts
- 3 clusters of 4 failures (1 in last day) out of 119245 builds

**Comment by carlory** [comment]
I can reproduce this issue on the master branch by the following steps:

```bash
(base) ➜  kubernetes git:(master) go test ./pkg/kubelet/volumemanager/reconciler -c -count=1 -race
(base) ➜  kubernetes git:(master) stress -p 300 ./reconciler.test -test.run "^Test_UncertainDeviceGlobalMounts/timed_out_operations_should_result_in_device_marked_as_uncertain_\[Block\]\[$"
5s: 0 runs so far, 0 failures
...
40s: 233 runs so far, 1 failures (0.43%)
…
7m5s: 3077 runs so far, 6 failures (0.19%)
...
14m55s: 6199 runs so far, 8 failures (0.13%)
15m0s: 6222 runs so far, 8 failures (0.13%)
15m5s: 6238 runs so far, 8 failures (0.13%)
15m10s: 6269 runs so far, 8 failures (0.13%)
15m15s: 6298 runs so far, 8 failures (0.13%)
15m20s: 6322 runs so far, 8 failures (0.13%)
```

And I can reproduce this issue on the v1.29.9 with the above steps.

```bash
(base) ➜  kubernetes git:(114a1f58037) go test ./pkg/kubelet/volumemanager/reconciler -c -count=1 -race
(base) ➜  kubernetes git:(114a1f58037) stress -p 300 ./reconciler.test -test.run "^Test_UncertainDeviceGlobalMounts/timed_out_operations_should_result_in_device_marked_as_uncertain_\[Block\]\[$"
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 0 runs so far, 0 failures
...
3m10s: 1211 runs so far, 0 failures
3m15s: 1238 runs so far, 1 failures (0.08%)
3m20s: 1284 runs so far, 1 failures (0.08%)
3m25s: 1302 runs so far, 1 failures (0.08%)
3m30s: 1329 runs so far, 1 failures (0.08%)
3m35s: 1354 runs so far, 1 failures (0.07%)
3m40s: 1395 runs so far, 1 failures (0.07%)
3m45s: 1440 runs so far, 1 failures (0.07%)
3m50s: 1462 runs so far, 1 failures (0.07%)
3m55s: 1484 runs so far, 1 failures (0.07%)
4m0s: 1531 runs so far, 1 failures (0.07%)
4m5s: 1593 runs so far, 1 failures (0.06%)
4m10s: 1624 runs so far, 1 failures (0.06%)
4m15s: 1653 runs so far, 1 failures (0.06%)
4m20s: 1718 runs so far, 1 failures (0.06%)
4m25s: 1758 runs so far, 1 failures (0.06%)
4m30s: 1779 runs so far, 1 failures (0.06%)
4m35s: 1805 runs so far, 1 failures (0.06%)
4m40s: 1840 runs so far, 1 failures (0.05%)
4m45s: 1868 runs so far, 1 failures (0.05%)
4m50s: 1903 runs so far, 1 failures (0.05%)
^C
```

I think it is an old flaky test. IMO, the test code is right. It should be a bug in the reconciler. Before detaching a device, the device should be teared down. It may be a critical bug. cc @jsafrane 

```
➜  ~ cat /var/folders/n6/33y75ytn36g38k2wnkb6lp8r0000gn/T/go-stress-20241017T191811-3448965742
E1017 19:21:21.839841    4992 reconstruct_new.go:56] "Cannot get volumes from disk, skip sync states for volume reconstruction" err="open fake-dir-6ec1e204152d0da11b247a1c9de569d6: no such file or directory"
I1017 19:21:21.920196    4992 reconciler_new.go:29] "Reconciler: start to sync state"
I1017 19:21:21.947373    4992 reconciler_common.go:258] "operationExecutor.VerifyControllerAttachedVolume started for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") " pod="pod1"
I1017 19:21:21.979791    4992 operation_generator.go:1579] "Controller attach succeeded for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") device path: \"fake/path\"" pod="pod1"
I1017 19:21:22.001027    4992 operation_generator.go:1099] "MapVolume.WaitForAttach entering for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") DevicePath \"fake/path\"" pod="pod1"
I1017 19:21:22.012754    4992 operation_generator.go:1109] "MapVolume.WaitForAttach succeeded for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") DevicePath \"/dev/sdb\"" pod="pod1"
E1017 19:21:22.040036    4992 nestedpendingoperations.go:348] Operation for "{volumeName:fake-plugin/timeout-mount-device-volume podName: nodeName:}" failed. No retries permitted until 2024-10-17 19:21:22.513028 +0800 CST m=+43.845587626 (durationBeforeRetry 500ms). Error: MapVolume.SetUpDevice failed for volume "timeout-mount-device-volume" (UniqueName: "fake-plugin/timeout-mount-device-volume") pod "pod1" (UID: "pod1uid") : mount failed
I1017 19:21:22.071487    4992 reconciler_common.go:300] "Volume detached for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") on node \"mynodename\" DevicePath \"fake/path\""
--- FAIL: Test_UncertainDeviceGlobalMounts (0.35s)
    --- FAIL: Test_UncertainDeviceGlobalMounts/timed_out_operations_should_result_in_device_marked_as_uncertain_[Block][ (0.74s)
        reconciler_test.go:1652: Error verifying UnMountDeviceCallCount: No Unmapper have expected TearDownDeviceCallCount. Expected: <1>.
FAIL


ERROR: exit status 1
```

In order to debug this issue, I will add some logs. From the logs, I can see that the device is marked as uncertain but when detaching the device, the mount state of the device is DeviceNotMounted, so the tear down operation is not executed. It is strange. I need more time to debug this issue and can't fix it in a short time. If someone can fix it, please help to fix it. If you have any ideas, please let me know.

```
E1016 18:59:24.678405   22368 reconstruct.go:54] "Cannot get volumes from disk, skip sync states for volume reconstruction" err="open fake-dir-6ec1e204152d0da11b247a1c9de569d6: no such file or directory"
I1016 18:59:24.678553   22368 reconciler.go:26] "Reconciler: start to sync state"
I1016 18:59:24.678651   22368 nestedpendingoperations.go:269] "Operation not exist" exist=false volumeName="fake-plugin/timeout-mount-device-volume" podName="" nodeName="" want="reconciler.MountOrAttachVolumes"
I1016 18:59:24.678752   22368 reconciler_common.go:251] "operationExecutor.VerifyControllerAttachedVolume started for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") " pod="pod1"
I1016 18:59:24.678839   22368 nestedpendingoperations.go:263] "Operation is pending" volumeName="fake-plugin/timeout-mount-device-volume" podName="" nodeName="" operationName="verify_controller_attached_volume" want="reconciler.MountOrAttachVolumes"
I1016 18:59:24.678885   22368 nestedpendingoperations.go:263] "Operation is pending" volumeName="fake-plugin/timeout-mount-device-volume" podName="" nodeName="" operationName="verify_controller_attached_volume" want="reconciler.MountOrAttachVolumes"
I1016 18:59:24.678922   22368 nestedpendingoperations.go:263] "Operation is pending" volumeName="fake-plugin/timeout-mount-device-volume" podName="" nodeName="" operationName="verify_controller_attached_volume" want="reconciler.MountOrAttachVolumes"
I1016 18:59:24.681712   22368 actual_state_of_world.go:390] "MarkVolumeAsAttached" volumeName="" devicePath="fake/path" pluginIsAttachable="True"
I1016 18:59:24.681875   22368 actual_state_of_world.go:706] "Volume is not added to attachedVolume list" volumeName="fake-plugin/timeout-mount-device-volume" path="fake/path" deviceMountState="DeviceNotMounted"
I1016 18:59:24.682624   22368 operation_generator.go:1492] "Controller attach succeeded for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") device path: \"fake/path\"" pod="pod1"
I1016 18:59:24.695466   22368 nestedpendingoperations.go:263] "Operation is pending" volumeName="fake-plugin/timeout-mount-device-volume" podName="" nodeName="" operationName="verify_controller_attached_volume" want="reconciler.MountOrAttachVolumes"
I1016 18:59:24.695679   22368 nestedpendingoperations.go:269] "Operation not exist" exist=false volumeName="fake-plugin/timeout-mount-device-volume" podName="" nodeName="" want="reconciler.MountOrAttachVolumes"
I1016 18:59:24.810149   22368 operation_generator.go:1014] "MapVolume.WaitForAttach entering for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") DevicePath \"fake/path\"" pod="pod1"
I1016 18:59:24.811509   22368 operation_generator.go:1024] "MapVolume.WaitForAttach succeeded for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") pod \"pod1\" (UID: \"pod1uid\") DevicePath \"/dev/sdb\"" pod="pod1"
I1016 18:59:24.811550   22368 operation_generator.go:1032] "debug GenerateMapVolumeFunc" globalMapPath="plugins/fake-plugin/volumeDevices/pluginDependentPath"
I1016 18:59:24.814429   22368 actual_state_of_world.go:559] "MarkDeviceAsUncertain" volumeName="fake-plugin/timeout-mount-device-volume" devicePath="/dev/sdb" deviceMountPath="plugins/fake-plugin/volumeDevices/pluginDependentPath"
I1016 18:59:24.846457   22368 actual_state_of_world.go:853] "SetDeviceMountState" volumeObj={} deviceMountState="DeviceMountUncertain" devicePath="/dev/sdb" deviceMountPath="plugins/fake-plugin/volumeDevices/pluginDependentPath"
E1016 18:59:24.871391   22368 nestedpendingoperations.go:383] Operation for "{volumeName:fake-plugin/timeout-mount-device-volume podName: nodeName:}" failed. No retries permitted until 2024-10-16 18:59:25.346757 +0800 CST m=+20.884467043 (durationBeforeRetry 500ms). Error: MapVolume.SetUpDevice failed for volume "timeout-mount-device-volume" (UniqueName: "fake-plugin/timeout-mount-device-volume") pod "pod1" (UID: "pod1uid") : mount failed
I1016 18:59:24.912845   22368 reconciler_test.go:1642] "dws.DeletePodFromVolume start"
I1016 18:59:24.926852   22368 reconciler_test.go:1644] "dws.DeletePodFromVolume end"
I1016 18:59:24.946856   22368 nestedpendingoperations.go:267] "Operation exist and not pending" exist=true volumeName="fake-plugin/timeout-mount-device-volume" podName="" nodeName="" operationName="map_volume" want="reconciler.UnmountDetachDevices"
I1016 18:59:25.627123   22368 reconciler_common.go:278] "unmountDetachDevices" attachedVolume={"VolumeName":"fake-plugin/timeout-mount-device-volume","VolumeSpec":{"Volume":null,"PersistentVolume":{"metadata":{"name":"timeout-mount-device-volume","uid":"pvuid","creationTimestamp":null},"spec":{"claimRef":{"name":"pvc"},"volumeMode":"Block"},"status":{}},"ReadOnly":false,"InlineVolumeSpecForCSIMigration":false,"Migrated":false},"NodeName":"mynodename","PluginIsAttachable":true,"DevicePath":"fake/path","DeviceMountPath":"","PluginName":"fake-plugin","DeviceMountState":"DeviceNotMounted","SELinuxMountContext":""}
I1016 18:59:25.638014   22368 reconciler_common.go:291] "unmountDetachDevices will mark Volume detached" attachedVolume={"VolumeName":"fake-plugin/timeout-mount-device-volume","VolumeSpec":{"Volume":null,"PersistentVolume":{"metadata":{"name":"timeout-mount-device-volume","uid":"pvuid","creationTimestamp":null},"spec":{"claimRef":{"name":"pvc"},"volumeMode":"Block"},"status":{}},"ReadOnly":false,"InlineVolumeSpecForCSIMigration":false,"Migrated":false},"NodeName":"mynodename","PluginIsAttachable":true,"DevicePath":"fake/path","DeviceMountPath":"","PluginName":"fake-plugin","DeviceMountState":"DeviceNotMounted","SELinuxMountContext":""} DeviceMayBeMounted="DeviceNotMounted"
I1016 18:59:25.656426   22368 reconciler_common.go:296] "Volume detached for volume \"timeout-mount-device-volume\" (UniqueName: \"fake-plugin/timeout-mount-device-volume\") on node \"mynodename\" DevicePath \"fake/path\""
--- FAIL: Test_UncertainDeviceGlobalMounts (0.02s)
    --- FAIL: Test_UncertainDeviceGlobalMounts/timed_out_operations_should_result_in_device_marked_as_uncertain_[Block][ (1.64s)
        reconciler_test.go:1656: Error verifying UnMountDeviceCallCount: No Unmapper have expected TearDownDeviceCallCount. Expected: <1>.
FAIL


ERROR: exit status 1
```

**Comment by drewhagen** [comment]
Hello @carlory @pacoxu  Thanks for taking action on this!

A friendly reminder of what's ahead:
* Code freeze is starting `02:00 UTC Friday November 8th 2024` (about **3 weeks** from now), and while there is still time, we want to ensure that each PR has a chance to be merged on time.

Given this timeline and capacity, will this continue to be aimed for the 1.32 release?
Thanks! 😄 🚀

**Comment by carlory** [comment]
> Given this timeline and capacity, will this continue to be aimed for the 1.32 release?

I am trying my best but I can not promise it
